casper/src/test/scala/coop/rchain/casper/genesis/DeployPaymentCostTest.scala

```diff
@@ -0,0 +1,257 @@
+    val captureTransferStatusChannel = GPrivateBuilder()
+    val transferStatus = node.runtimeManager.captureResults(
+      paymentBlock.getBody.getPostState.tuplespace,
+      captureSingleResult(GString(statusChannel), captureTransferStatusChannel),
+      captureTransferStatusChannel
+    )
+    assert(transferStatus.size == 1)
+    assert(transferStatus.head.exprs.head == Expr(GString("Success")))
+
+    println(s"Cost of deploying payment contract is at minimum: ${paymentDeployCost.head.cost}")
+  }
+
+}
+
+object DeployPaymentCostTest {
+  val pubKeyString: String = "11afb9a5fa2b3e194b701987b3531a93dbdf790dac26f8a2502cfa5d529f6b4d"
+  val secKeyString: String = "d039d5c634ad95d968fc18368d81b97aaecd32fc7cf6eec07a97c5ac9f9fcb5b"
+
+  import HashSetCasperTest._
+
+  private val (validatorKeys, validators) = (1 to 4).map(_ => Ed25519.newKeyPair).unzip
+  private val (_, ethPubKeys)             = (1 to 4).map(_ => Secp256k1.newKeyPair).unzip
+  private val ethAddresses =
+    ethPubKeys.map(pk => "0x" + Base16.encode(Keccak256.hash(pk.bytes.drop(1)).takeRight(20)))
+  private val wallets     = ethAddresses.map(addr => PreWallet(addr, BigInt(10001)))
+  private val bonds       = createBonds(validators)
+  private val minimumBond = 100L
+  private val genesis =
+    buildGenesis(wallets, bonds, minimumBond, Long.MaxValue, Faucet.basicWalletFaucet, 0L)
+
+  def bytes(par: Par): Array[Byte] = {
+    import coop.rchain.models.serialization.implicits._
+    Serialize[Par].encode(par).toArray
+  }
+
+  def deploy[F[_]: Apply](
+      dd: DeployData
+  )(implicit casperEff: MultiParentCasperImpl[F]): BlockMessage = {
+    val Created(block) = casperEff.deploy(dd) *> casperEff.createBlock
+    val blockStatus    = casperEff.addBlock(block)
+    assert(blockStatus == Valid)
+    block
+  }
+
+  def deploy[F[_]: Apply](
+      deploy: Deploy
+  )(implicit casperEff: MultiParentCasperImpl[F]): BlockMessage = {
+    val Created(block) = casperEff.deploy(deploy) *> casperEff.createBlock
+    val blockStatus    = casperEff.addBlock(block)
+    assert(blockStatus == Valid)
+    block
+  }
+
+  def deploy[F[_]: Apply](par: Par)(implicit casperEff: MultiParentCasperImpl[F]): BlockMessage = {
+    val parDeploy      = ProtoUtil.termDeployNow(par)
+    val Created(block) = casperEff.deploy(parDeploy) *> casperEff.createBlock
+    val blockStatus    = casperEff.addBlock(block)
+    assert(blockStatus == Valid)
+    block
+  }
+
+  def deployAndCapture(p: Par, retChannel: Par, rm: RuntimeManager)(
+      implicit casper: MultiParentCasperImpl[Id]
+  ): Par = {
+    val postDeployStateHash = deploy[Id](p).getBody.getPostState.tuplespace
+    val captureChan         = GPrivateBuilder()
+    val res = rm.captureResults(
+      postDeployStateHash,
+      captureSingleResult(retChannel, captureChan),
+      captureChan
+    )
+    assert(res.size == 1)
+    res.head
+  }
+
+  def registerSigned(tuple: Par, pubKey: ByteString, sig: ByteString, retChannel: Par): Par =
```

Comment:
Thank you for extracting this and the next method :)

I also think some kind of a macro language would help us greatly. e.g.

```scala
interpolate("""
  new x(`rho:registry:insertSigned:ed25519`) {
    x!(p0, p1, p2, p3)
  }
""", GByteArray(pubKey), tuple, GByteArray(sig), retChannel)

def interpolate(rho: String, pars: Par*) = ???
```

For a start, we would only support interpolating Par-s (vs. Exprs and other production types). The `p1...pn` substitution could be as easy as doing the following in the `interpolate`:

```scala
def interpolate(rho: String, pars: Par*) = {
  val template = s"""
    for (@p1, @p2, @p3, @p4 <- @0) {
      $rho
    }
  """
  //then, I think, we only need to do one step of substitution
  //and we can extract the $rho substituted with the parameters and return it
}
```

What do you think?

Reply:
I have been thinking about some kind of interpolation for the rholang when I was on holidays but it would work bit differently than in your example. Anyway - this is not something that should be done in this PR. I will create a JIRA for that.

Reply:
>  `    for (@p1, @p2, @p3, @p4 <- @0) {`

this will become a noop once we remove sending/receiving on forgeable names.